{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load and Normalize Data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats as stats"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create DataFrame from TEP Count Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5440, 2351)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1-Vumc-HD-101-TR922</th>\n",
       "      <th>2-Vumc-HD-103-TR923</th>\n",
       "      <th>3-Vumc-HD-108-TR924</th>\n",
       "      <th>4-Vumc-HD-127-TR925</th>\n",
       "      <th>5-Vumc-HD-130-TR926</th>\n",
       "      <th>6-Vumc-HD-142-TR927</th>\n",
       "      <th>7-Vumc-HD-143-TR928</th>\n",
       "      <th>8-Vumc-HD-144-TR929</th>\n",
       "      <th>9-Vumc-HD-145-TR930</th>\n",
       "      <th>11-Vumc-HD-149-TR932</th>\n",
       "      <th>...</th>\n",
       "      <th>countMatrix.3502-NKI-4854-SARC-TR3080</th>\n",
       "      <th>countMatrix.3503-NKI-4707-SARC-TR3079</th>\n",
       "      <th>countMatrix.3504-NKI-4941-SARC-TR3084</th>\n",
       "      <th>countMatrix.3507-NKI-6348-SARC-TR3134</th>\n",
       "      <th>countMatrix.3508-NKI-5275-SARC-TR3099</th>\n",
       "      <th>countMatrix.3512-NKI-5148-SARC-TR3129</th>\n",
       "      <th>countMatrix.320-10</th>\n",
       "      <th>countMatrix.327-34</th>\n",
       "      <th>countMatrix.376-P-PH-001-013-TR1400</th>\n",
       "      <th>countMatrix.506-NKI-NSCLC-107-270</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>5</td>\n",
       "      <td>79</td>\n",
       "      <td>58</td>\n",
       "      <td>128</td>\n",
       "      <td>93</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>42</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>63</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>24</td>\n",
       "      <td>336</td>\n",
       "      <td>334</td>\n",
       "      <td>59</td>\n",
       "      <td>120</td>\n",
       "      <td>22</td>\n",
       "      <td>31</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>163</td>\n",
       "      <td>90</td>\n",
       "      <td>46</td>\n",
       "      <td>75</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <td>38</td>\n",
       "      <td>46</td>\n",
       "      <td>15</td>\n",
       "      <td>36</td>\n",
       "      <td>25</td>\n",
       "      <td>43</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>59</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001461</th>\n",
       "      <td>72</td>\n",
       "      <td>17</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>49</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2351 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 1-Vumc-HD-101-TR922  2-Vumc-HD-103-TR923  \\\n",
       "ENSG00000000419                    5                   79   \n",
       "ENSG00000000460                    8                   10   \n",
       "ENSG00000000938                   24                  336   \n",
       "ENSG00000001036                   38                   46   \n",
       "ENSG00000001461                   72                   17   \n",
       "\n",
       "                 3-Vumc-HD-108-TR924  4-Vumc-HD-127-TR925  \\\n",
       "ENSG00000000419                   58                  128   \n",
       "ENSG00000000460                   21                   14   \n",
       "ENSG00000000938                  334                   59   \n",
       "ENSG00000001036                   15                   36   \n",
       "ENSG00000001461                   11                    9   \n",
       "\n",
       "                 5-Vumc-HD-130-TR926  6-Vumc-HD-142-TR927  \\\n",
       "ENSG00000000419                   93                   12   \n",
       "ENSG00000000460                    6                    8   \n",
       "ENSG00000000938                  120                   22   \n",
       "ENSG00000001036                   25                   43   \n",
       "ENSG00000001461                   18                   28   \n",
       "\n",
       "                 7-Vumc-HD-143-TR928  8-Vumc-HD-144-TR929  \\\n",
       "ENSG00000000419                    0                    0   \n",
       "ENSG00000000460                   25                   42   \n",
       "ENSG00000000938                   31                   13   \n",
       "ENSG00000001036                   27                   27   \n",
       "ENSG00000001461                    5                    6   \n",
       "\n",
       "                 9-Vumc-HD-145-TR930  11-Vumc-HD-149-TR932  ...  \\\n",
       "ENSG00000000419                   18                    28  ...   \n",
       "ENSG00000000460                    5                     7  ...   \n",
       "ENSG00000000938                   19                    36  ...   \n",
       "ENSG00000001036                   59                     5  ...   \n",
       "ENSG00000001461                   49                    18  ...   \n",
       "\n",
       "                 countMatrix.3502-NKI-4854-SARC-TR3080  \\\n",
       "ENSG00000000419                                      3   \n",
       "ENSG00000000460                                      0   \n",
       "ENSG00000000938                                      8   \n",
       "ENSG00000001036                                      1   \n",
       "ENSG00000001461                                      1   \n",
       "\n",
       "                 countMatrix.3503-NKI-4707-SARC-TR3079  \\\n",
       "ENSG00000000419                                      0   \n",
       "ENSG00000000460                                      3   \n",
       "ENSG00000000938                                     45   \n",
       "ENSG00000001036                                      3   \n",
       "ENSG00000001461                                      1   \n",
       "\n",
       "                 countMatrix.3504-NKI-4941-SARC-TR3084  \\\n",
       "ENSG00000000419                                      1   \n",
       "ENSG00000000460                                      0   \n",
       "ENSG00000000938                                      1   \n",
       "ENSG00000001036                                      4   \n",
       "ENSG00000001461                                      0   \n",
       "\n",
       "                 countMatrix.3507-NKI-6348-SARC-TR3134  \\\n",
       "ENSG00000000419                                      2   \n",
       "ENSG00000000460                                      2   \n",
       "ENSG00000000938                                      7   \n",
       "ENSG00000001036                                     15   \n",
       "ENSG00000001461                                      0   \n",
       "\n",
       "                 countMatrix.3508-NKI-5275-SARC-TR3099  \\\n",
       "ENSG00000000419                                      9   \n",
       "ENSG00000000460                                     63   \n",
       "ENSG00000000938                                    163   \n",
       "ENSG00000001036                                      3   \n",
       "ENSG00000001461                                     12   \n",
       "\n",
       "                 countMatrix.3512-NKI-5148-SARC-TR3129  countMatrix.320-10  \\\n",
       "ENSG00000000419                                     16                  22   \n",
       "ENSG00000000460                                      3                   3   \n",
       "ENSG00000000938                                     90                  46   \n",
       "ENSG00000001036                                      6                  10   \n",
       "ENSG00000001461                                      2                   6   \n",
       "\n",
       "                 countMatrix.327-34  countMatrix.376-P-PH-001-013-TR1400  \\\n",
       "ENSG00000000419                  21                                    0   \n",
       "ENSG00000000460                   0                                    1   \n",
       "ENSG00000000938                  75                                   14   \n",
       "ENSG00000001036                   8                                    3   \n",
       "ENSG00000001461                   6                                    1   \n",
       "\n",
       "                 countMatrix.506-NKI-NSCLC-107-270  \n",
       "ENSG00000000419                                  1  \n",
       "ENSG00000000460                                  0  \n",
       "ENSG00000000938                                  8  \n",
       "ENSG00000001036                                  0  \n",
       "ENSG00000001461                                  0  \n",
       "\n",
       "[5 rows x 2351 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read data file into a panda dataframe\n",
    "file_in = 'TEP_Count_Matrix.tsv.gz'\n",
    "df_data = pd.read_csv(file_in, sep='\\t')\n",
    "print(df_data.shape)   # 5440 genes, 2351 samples\n",
    "df_data.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000000419', 'ENSG00000000460', 'ENSG00000000938',\n",
       "       'ENSG00000001036', 'ENSG00000001461', 'ENSG00000001497',\n",
       "       'ENSG00000001629', 'ENSG00000001631', 'ENSG00000002330',\n",
       "       'ENSG00000002549',\n",
       "       ...\n",
       "       'ENSG00000267279', 'ENSG00000267316', 'ENSG00000268555',\n",
       "       'ENSG00000269028', 'ENSG00000270055', 'ENSG00000271043',\n",
       "       'ENSG00000272053', 'ENSG00000272168', 'ENSG00000272369',\n",
       "       'ENSG00000272888'],\n",
       "      dtype='object', length=5440)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_genes = df_data.index\n",
    "list_of_genes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalize TEP Count Matrix"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look for gene length in GFF3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(196520, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000456328</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>1657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000515242</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>1653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000518655</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>1483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000438504</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>1783</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            gene  Length\n",
       "ENST00000456328  ENSG00000223972    1657\n",
       "ENST00000515242  ENSG00000223972    1653\n",
       "ENST00000518655  ENSG00000223972    1483\n",
       "ENST00000450305  ENSG00000223972     632\n",
       "ENST00000438504  ENSG00000227232    1783"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read GFF3\n",
    "file='GFF3.csv'\n",
    "df_GFF3 = pd.read_csv(file,sep=';',na_values='n.a.',index_col=0)\n",
    "print(df_GFF3.shape)\n",
    "df_GFF3.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>number_of_transcripts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>57820.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>3.398824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>4.880759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>82.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       number_of_transcripts\n",
       "count           57820.000000\n",
       "mean                3.398824\n",
       "std                 4.880759\n",
       "min                 1.000000\n",
       "25%                 1.000000\n",
       "50%                 1.000000\n",
       "75%                 4.000000\n",
       "max                82.000000"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check how many transcripts we have per gene\n",
    "df_GFF3_transcripts_per_gene = df_GFF3.groupby('gene').count()\n",
    "df_GFF3_transcripts_per_gene = df_GFF3_transcripts_per_gene.rename(columns = {'Length':'number_of_transcripts'})\n",
    "df_GFF3_transcripts_per_gene.describe()\n",
    "# df_GFF3_transcripts_per_gene.head(3)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each gene has between 1 and 82 transcripts. \n",
    "\n",
    "In the pacients database we only have the name of the gene, but not the name of the transcript measured. Therefore, we need to reduce number of transcripts per gene to have a more accurate Length number to use in the normalization."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look for gene length finding most common isoform per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size:  (25508, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>uc010nxq.1</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>1483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>uc009viu.3</td>\n",
       "      <td>ENST00000438504</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>1783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>uc001aak.3</td>\n",
       "      <td>ENST00000417324</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>1187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>uc001aal.1</td>\n",
       "      <td>ENST00000335137</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>uc021oeg.2</td>\n",
       "      <td>ENST00000493797</td>\n",
       "      <td>ENSG00000239906</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         name       transcript             gene  Length\n",
       "0  uc010nxq.1  ENST00000518655  ENSG00000223972    1483\n",
       "1  uc009viu.3  ENST00000438504  ENSG00000227232    1783\n",
       "2  uc001aak.3  ENST00000417324  ENSG00000237613    1187\n",
       "3  uc001aal.1  ENST00000335137  ENSG00000186092     918\n",
       "4  uc021oeg.2  ENST00000493797  ENSG00000239906     323"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load Best Transcript file\n",
    "file='BestTranscript.csv'\n",
    "df_best_transcripts = pd.read_csv(file, sep=';')\n",
    "print(\"Size: \", df_best_transcripts.shape)\n",
    "df_best_transcripts.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 3 cases\n",
    "# a) Genes that appear only once in df_best_transcripts\n",
    "# b) Genes that appear MORE THAN once in df_best_transcripts\n",
    "# c) Genes that DO NOT appear in df_best_transcripts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5454, 4)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a dataframe with those best transcripts whose gene appear in the patients dataframe\n",
    "df_best_transcripts_subset = df_best_transcripts[df_best_transcripts['gene'].isin(list_of_genes)]\n",
    "df_best_transcripts_subset.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size:  (5251, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>uc001abz.4</td>\n",
       "      <td>ENST00000327044</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>2790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>uc001adh.4</td>\n",
       "      <td>ENST00000263741</td>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>uc001ado.3</td>\n",
       "      <td>ENST00000450390</td>\n",
       "      <td>ENSG00000160087</td>\n",
       "      <td>2455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>uc001aeb.2</td>\n",
       "      <td>ENST00000354700</td>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>uc001aee.2</td>\n",
       "      <td>ENST00000435064</td>\n",
       "      <td>ENSG00000127054</td>\n",
       "      <td>2148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          name       transcript             gene  Length\n",
       "16  uc001abz.4  ENST00000327044  ENSG00000188976    2790\n",
       "34  uc001adh.4  ENST00000263741  ENSG00000078808    2079\n",
       "37  uc001ado.3  ENST00000450390  ENSG00000160087    2455\n",
       "39  uc001aeb.2  ENST00000354700  ENSG00000131584    3896\n",
       "41  uc001aee.2  ENST00000435064  ENSG00000127054    2148"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a) Genes that appear only once in df_best_transcripts\n",
    "df_best_transcripts_subset_once = df_best_transcripts_subset.drop_duplicates(subset=['gene'], keep=False, inplace=False)\n",
    "\n",
    "print(\"Size: \", df_best_transcripts_subset_once.shape)\n",
    "df_best_transcripts_subset_once.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>2790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENSG00000160087</td>\n",
       "      <td>2455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ENSG00000131584</td>\n",
       "      <td>3896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>ENSG00000127054</td>\n",
       "      <td>2148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene  Length\n",
       "16  ENSG00000188976    2790\n",
       "34  ENSG00000078808    2079\n",
       "37  ENSG00000160087    2455\n",
       "39  ENSG00000131584    3896\n",
       "41  ENSG00000127054    2148"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_best_transcripts_subset_once_red = df_best_transcripts_subset_once.drop(['name', 'transcript'], axis = 1, inplace = False)\n",
    "df_best_transcripts_subset_once_red.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes that appear only once in best transcripts database represent  96.52573529411765 % of genes\n"
     ]
    }
   ],
   "source": [
    "print(\"Genes that appear only once in best transcripts database represent \", 5251*100/5440, \"% of genes\")\n",
    "# Only 3.5% of genes are not completely defined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size:  (203, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>uc001ajt.1</td>\n",
       "      <td>ENST00000463471</td>\n",
       "      <td>ENSG00000157873</td>\n",
       "      <td>4255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>uc001ajr.3</td>\n",
       "      <td>ENST00000355716</td>\n",
       "      <td>ENSG00000157873</td>\n",
       "      <td>1707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>uc009vtt.3</td>\n",
       "      <td>ENST00000373720</td>\n",
       "      <td>ENSG00000060688</td>\n",
       "      <td>1087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>uc001bso.3</td>\n",
       "      <td>ENST00000263694</td>\n",
       "      <td>ENSG00000060688</td>\n",
       "      <td>1595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>uc001byq.3</td>\n",
       "      <td>ENST00000468598</td>\n",
       "      <td>ENSG00000116560</td>\n",
       "      <td>2835</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           name       transcript             gene  Length\n",
       "90   uc001ajt.1  ENST00000463471  ENSG00000157873    4255\n",
       "91   uc001ajr.3  ENST00000355716  ENSG00000157873    1707\n",
       "502  uc009vtt.3  ENST00000373720  ENSG00000060688    1087\n",
       "503  uc001bso.3  ENST00000263694  ENSG00000060688    1595\n",
       "571  uc001byq.3  ENST00000468598  ENSG00000116560    2835"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# b) Genes that appear MORE THAN once in df_best_transcripts\n",
    "df_best_transcripts_subset_more_than_once = df_best_transcripts_subset[df_best_transcripts_subset.duplicated(subset=['gene'], keep=False)]\n",
    "\n",
    "print(\"Size: \", df_best_transcripts_subset_more_than_once.shape)\n",
    "df_best_transcripts_subset_more_than_once.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size:  (98, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>transcript</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000007392</th>\n",
       "      <td>uc010bqq.1</td>\n",
       "      <td>ENST00000490762</td>\n",
       "      <td>2643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000011275</th>\n",
       "      <td>uc003spb.3</td>\n",
       "      <td>ENST00000476345</td>\n",
       "      <td>5639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000011454</th>\n",
       "      <td>uc011lzk.3</td>\n",
       "      <td>ENST00000468760</td>\n",
       "      <td>4997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000026297</th>\n",
       "      <td>uc003qvi.1</td>\n",
       "      <td>ENST00000508775</td>\n",
       "      <td>2445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000033050</th>\n",
       "      <td>uc031szt.1</td>\n",
       "      <td>ENST00000287844</td>\n",
       "      <td>3546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       name       transcript  Length\n",
       "gene                                                \n",
       "ENSG00000007392  uc010bqq.1  ENST00000490762    2643\n",
       "ENSG00000011275  uc003spb.3  ENST00000476345    5639\n",
       "ENSG00000011454  uc011lzk.3  ENST00000468760    4997\n",
       "ENSG00000026297  uc003qvi.1  ENST00000508775    2445\n",
       "ENSG00000033050  uc031szt.1  ENST00000287844    3546"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For each gene, we have to choose which transcript length to use.\n",
    "# In this case, we are going to select the maximum length between those transcripts that appear in Best_transcripts\n",
    "# Before applying normalization, we will decide if we want to use these genes or not\n",
    "df_best_transcripts_subset_more_than_once = df_best_transcripts_subset_more_than_once.groupby('gene').max()\n",
    "print(\"Size: \", df_best_transcripts_subset_more_than_once.shape)\n",
    "df_best_transcripts_subset_more_than_once.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "98"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_genes_more_than_once = list(df_best_transcripts_subset_more_than_once.index)\n",
    "len(list_of_genes_more_than_once)\n",
    "# list_of_genes_more_than_once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000007392</td>\n",
       "      <td>2643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000011275</td>\n",
       "      <td>5639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000011454</td>\n",
       "      <td>4997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000026297</td>\n",
       "      <td>2445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000033050</td>\n",
       "      <td>3546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene  Length\n",
       "0  ENSG00000007392    2643\n",
       "1  ENSG00000011275    5639\n",
       "2  ENSG00000011454    4997\n",
       "3  ENSG00000026297    2445\n",
       "4  ENSG00000033050    3546"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_best_transcripts_subset_more_than_once_red = df_best_transcripts_subset_more_than_once.drop(['name', 'transcript'], axis = 1, inplace = False)\n",
    "df_best_transcripts_subset_more_than_once_red.reset_index(inplace=True)\n",
    "df_best_transcripts_subset_more_than_once_red.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# c) Genes that DO NOT appear in df_best_transcripts\n",
    "list_of_best_transcripts_genes = list(df_best_transcripts['gene'])\n",
    "len(list_of_best_transcripts_genes)\n",
    "list_of_genes_not_appear = [x for x in list_of_genes if x not in list_of_best_transcripts_genes]\n",
    "len(list_of_genes_not_appear)\n",
    "# list_of_genes_not_appear"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size:  (91, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000066697</th>\n",
       "      <td>1991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000104823</th>\n",
       "      <td>1411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000113575</th>\n",
       "      <td>4649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000128383</th>\n",
       "      <td>1478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000136783</th>\n",
       "      <td>1633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Length\n",
       "gene                   \n",
       "ENSG00000066697    1991\n",
       "ENSG00000104823    1411\n",
       "ENSG00000113575    4649\n",
       "ENSG00000128383    1478\n",
       "ENSG00000136783    1633"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For those genes, we have to choose which transcript length to use.\n",
    "# In this case, we are going to use the maximum length between those transcripts that appear in GFF3\n",
    "# Before applying normalization, we will decide if we want to use these genes or not\n",
    "df_best_transcripts_subset_not_appear = df_GFF3[df_GFF3['gene'].isin(list_of_genes_not_appear)]\n",
    "df_best_transcripts_subset_not_appear = df_best_transcripts_subset_not_appear.groupby('gene').max()\n",
    "\n",
    "print(\"Size: \", df_best_transcripts_subset_not_appear.shape)\n",
    "df_best_transcripts_subset_not_appear.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000066697</td>\n",
       "      <td>1991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000104823</td>\n",
       "      <td>1411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000113575</td>\n",
       "      <td>4649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000128383</td>\n",
       "      <td>1478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000136783</td>\n",
       "      <td>1633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene  Length\n",
       "0  ENSG00000066697    1991\n",
       "1  ENSG00000104823    1411\n",
       "2  ENSG00000113575    4649\n",
       "3  ENSG00000128383    1478\n",
       "4  ENSG00000136783    1633"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_best_transcripts_subset_not_appear_red = df_best_transcripts_subset_not_appear.reset_index()\n",
    "df_best_transcripts_subset_not_appear_red.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quick check\n",
    "# Gene ENSG00000066697\n",
    "# From gencode.v19.annotation it has the following transcripts:\n",
    "# transcript\tID=ENST00000395067\n",
    "# transcript\tID=ENST00000398977\n",
    "# transcript\tID=ENST00000374885\n",
    "# transcript\tID=ENST00000374886\n",
    "# transcript\tID=ENST00000489377\n",
    "\n",
    "# From knowToEnsemble\n",
    "# transcript\tID=ENST00000395067      uc022bkz\n",
    "# transcript\tID=ENST00000395067      uc004baw\n",
    "# transcript\tID=ENST00000395067      uc022bla\n",
    "# transcript\tID=ENST00000395067      uc022blb\n",
    "\n",
    "# transcript\tID=ENST00000398977      Not Found\n",
    "# transcript\tID=ENST00000374885      Not Found\n",
    "# transcript\tID=ENST00000374886      Not Found\n",
    "\n",
    "# transcript\tID=ENST00000489377      uc004bax\n",
    "\n",
    "# From knownCanonical\n",
    "# transcript\tID=ENST00000395067      uc022bkz       Not Found  \n",
    "# transcript\tID=ENST00000395067      uc004baw       Not Found\n",
    "# transcript\tID=ENST00000395067      uc022bla       Not Found\n",
    "# transcript\tID=ENST00000395067      uc022blb       Not Found\n",
    "\n",
    "# transcript\tID=ENST00000489377      uc004bax       Not Found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size:  (5251, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>2790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENSG00000078808</td>\n",
       "      <td>2079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENSG00000160087</td>\n",
       "      <td>2455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene  Length\n",
       "16  ENSG00000188976    2790\n",
       "34  ENSG00000078808    2079\n",
       "37  ENSG00000160087    2455"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a unique dataframe with all genes we want to use and lengths\n",
    "# In this case, we keep only those genes that have a best transcript\n",
    "# df_genes_length = pd.concat([df_best_transcripts_subset_once_red,df_best_transcripts_subset_more_than_once_red,df_best_transcripts_subset_not_appear_red], axis=0)\n",
    "df_genes_length = df_best_transcripts_subset_once_red.copy(deep=True)\n",
    "print(\"Size: \", df_genes_length.shape)  # 5440 - 91 - 98 = 5251\n",
    "df_genes_length.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the complete list of genes that appeared more than once or zero times in best transcripts list\n",
    "# And save it\n",
    "list_of_genes_to_check = list_of_genes_not_appear + list_of_genes_more_than_once\n",
    "len(list_of_genes_to_check)\n",
    "df = pd.DataFrame(list_of_genes_to_check)\n",
    "df.to_csv('List_of_genes_to_check.csv', index=False, header=None)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Apply Normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene      False\n",
      "Length    False\n",
      "dtype: bool\n"
     ]
    }
   ],
   "source": [
    "# Verify there is no transcript with length zero\n",
    "print((df_genes_length == 0).all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5440, 2351)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1-Vumc-HD-101-TR922</th>\n",
       "      <th>2-Vumc-HD-103-TR923</th>\n",
       "      <th>3-Vumc-HD-108-TR924</th>\n",
       "      <th>4-Vumc-HD-127-TR925</th>\n",
       "      <th>5-Vumc-HD-130-TR926</th>\n",
       "      <th>6-Vumc-HD-142-TR927</th>\n",
       "      <th>7-Vumc-HD-143-TR928</th>\n",
       "      <th>8-Vumc-HD-144-TR929</th>\n",
       "      <th>9-Vumc-HD-145-TR930</th>\n",
       "      <th>11-Vumc-HD-149-TR932</th>\n",
       "      <th>...</th>\n",
       "      <th>countMatrix.3502-NKI-4854-SARC-TR3080</th>\n",
       "      <th>countMatrix.3503-NKI-4707-SARC-TR3079</th>\n",
       "      <th>countMatrix.3504-NKI-4941-SARC-TR3084</th>\n",
       "      <th>countMatrix.3507-NKI-6348-SARC-TR3134</th>\n",
       "      <th>countMatrix.3508-NKI-5275-SARC-TR3099</th>\n",
       "      <th>countMatrix.3512-NKI-5148-SARC-TR3129</th>\n",
       "      <th>countMatrix.320-10</th>\n",
       "      <th>countMatrix.327-34</th>\n",
       "      <th>countMatrix.376-P-PH-001-013-TR1400</th>\n",
       "      <th>countMatrix.506-NKI-NSCLC-107-270</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>5</td>\n",
       "      <td>79</td>\n",
       "      <td>58</td>\n",
       "      <td>128</td>\n",
       "      <td>93</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>42</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>63</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>24</td>\n",
       "      <td>336</td>\n",
       "      <td>334</td>\n",
       "      <td>59</td>\n",
       "      <td>120</td>\n",
       "      <td>22</td>\n",
       "      <td>31</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>163</td>\n",
       "      <td>90</td>\n",
       "      <td>46</td>\n",
       "      <td>75</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 2351 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 1-Vumc-HD-101-TR922  2-Vumc-HD-103-TR923  \\\n",
       "ENSG00000000419                    5                   79   \n",
       "ENSG00000000460                    8                   10   \n",
       "ENSG00000000938                   24                  336   \n",
       "\n",
       "                 3-Vumc-HD-108-TR924  4-Vumc-HD-127-TR925  \\\n",
       "ENSG00000000419                   58                  128   \n",
       "ENSG00000000460                   21                   14   \n",
       "ENSG00000000938                  334                   59   \n",
       "\n",
       "                 5-Vumc-HD-130-TR926  6-Vumc-HD-142-TR927  \\\n",
       "ENSG00000000419                   93                   12   \n",
       "ENSG00000000460                    6                    8   \n",
       "ENSG00000000938                  120                   22   \n",
       "\n",
       "                 7-Vumc-HD-143-TR928  8-Vumc-HD-144-TR929  \\\n",
       "ENSG00000000419                    0                    0   \n",
       "ENSG00000000460                   25                   42   \n",
       "ENSG00000000938                   31                   13   \n",
       "\n",
       "                 9-Vumc-HD-145-TR930  11-Vumc-HD-149-TR932  ...  \\\n",
       "ENSG00000000419                   18                    28  ...   \n",
       "ENSG00000000460                    5                     7  ...   \n",
       "ENSG00000000938                   19                    36  ...   \n",
       "\n",
       "                 countMatrix.3502-NKI-4854-SARC-TR3080  \\\n",
       "ENSG00000000419                                      3   \n",
       "ENSG00000000460                                      0   \n",
       "ENSG00000000938                                      8   \n",
       "\n",
       "                 countMatrix.3503-NKI-4707-SARC-TR3079  \\\n",
       "ENSG00000000419                                      0   \n",
       "ENSG00000000460                                      3   \n",
       "ENSG00000000938                                     45   \n",
       "\n",
       "                 countMatrix.3504-NKI-4941-SARC-TR3084  \\\n",
       "ENSG00000000419                                      1   \n",
       "ENSG00000000460                                      0   \n",
       "ENSG00000000938                                      1   \n",
       "\n",
       "                 countMatrix.3507-NKI-6348-SARC-TR3134  \\\n",
       "ENSG00000000419                                      2   \n",
       "ENSG00000000460                                      2   \n",
       "ENSG00000000938                                      7   \n",
       "\n",
       "                 countMatrix.3508-NKI-5275-SARC-TR3099  \\\n",
       "ENSG00000000419                                      9   \n",
       "ENSG00000000460                                     63   \n",
       "ENSG00000000938                                    163   \n",
       "\n",
       "                 countMatrix.3512-NKI-5148-SARC-TR3129  countMatrix.320-10  \\\n",
       "ENSG00000000419                                     16                  22   \n",
       "ENSG00000000460                                      3                   3   \n",
       "ENSG00000000938                                     90                  46   \n",
       "\n",
       "                 countMatrix.327-34  countMatrix.376-P-PH-001-013-TR1400  \\\n",
       "ENSG00000000419                  21                                    0   \n",
       "ENSG00000000460                   0                                    1   \n",
       "ENSG00000000938                  75                                   14   \n",
       "\n",
       "                 countMatrix.506-NKI-NSCLC-107-270  \n",
       "ENSG00000000419                                  1  \n",
       "ENSG00000000460                                  0  \n",
       "ENSG00000000938                                  8  \n",
       "\n",
       "[3 rows x 2351 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <td>2790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000078808</th>\n",
       "      <td>2079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160087</th>\n",
       "      <td>2455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Length\n",
       "gene                   \n",
       "ENSG00000188976    2790\n",
       "ENSG00000078808    2079\n",
       "ENSG00000160087    2455"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes_length.index=df_genes_length['gene']\n",
    "df_genes_length.drop(['gene'], axis=1, inplace=True)\n",
    "df_genes_length.head(3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size:  (5251, 2352)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1-Vumc-HD-101-TR922</th>\n",
       "      <th>2-Vumc-HD-103-TR923</th>\n",
       "      <th>3-Vumc-HD-108-TR924</th>\n",
       "      <th>4-Vumc-HD-127-TR925</th>\n",
       "      <th>5-Vumc-HD-130-TR926</th>\n",
       "      <th>6-Vumc-HD-142-TR927</th>\n",
       "      <th>7-Vumc-HD-143-TR928</th>\n",
       "      <th>8-Vumc-HD-144-TR929</th>\n",
       "      <th>9-Vumc-HD-145-TR930</th>\n",
       "      <th>11-Vumc-HD-149-TR932</th>\n",
       "      <th>...</th>\n",
       "      <th>countMatrix.3503-NKI-4707-SARC-TR3079</th>\n",
       "      <th>countMatrix.3504-NKI-4941-SARC-TR3084</th>\n",
       "      <th>countMatrix.3507-NKI-6348-SARC-TR3134</th>\n",
       "      <th>countMatrix.3508-NKI-5275-SARC-TR3099</th>\n",
       "      <th>countMatrix.3512-NKI-5148-SARC-TR3129</th>\n",
       "      <th>countMatrix.320-10</th>\n",
       "      <th>countMatrix.327-34</th>\n",
       "      <th>countMatrix.376-P-PH-001-013-TR1400</th>\n",
       "      <th>countMatrix.506-NKI-NSCLC-107-270</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>5</td>\n",
       "      <td>79</td>\n",
       "      <td>58</td>\n",
       "      <td>128</td>\n",
       "      <td>93</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>42</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>63</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>24</td>\n",
       "      <td>336</td>\n",
       "      <td>334</td>\n",
       "      <td>59</td>\n",
       "      <td>120</td>\n",
       "      <td>22</td>\n",
       "      <td>31</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>163</td>\n",
       "      <td>90</td>\n",
       "      <td>46</td>\n",
       "      <td>75</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>2729</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 2352 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 1-Vumc-HD-101-TR922  2-Vumc-HD-103-TR923  \\\n",
       "ENSG00000000419                    5                   79   \n",
       "ENSG00000000460                    8                   10   \n",
       "ENSG00000000938                   24                  336   \n",
       "\n",
       "                 3-Vumc-HD-108-TR924  4-Vumc-HD-127-TR925  \\\n",
       "ENSG00000000419                   58                  128   \n",
       "ENSG00000000460                   21                   14   \n",
       "ENSG00000000938                  334                   59   \n",
       "\n",
       "                 5-Vumc-HD-130-TR926  6-Vumc-HD-142-TR927  \\\n",
       "ENSG00000000419                   93                   12   \n",
       "ENSG00000000460                    6                    8   \n",
       "ENSG00000000938                  120                   22   \n",
       "\n",
       "                 7-Vumc-HD-143-TR928  8-Vumc-HD-144-TR929  \\\n",
       "ENSG00000000419                    0                    0   \n",
       "ENSG00000000460                   25                   42   \n",
       "ENSG00000000938                   31                   13   \n",
       "\n",
       "                 9-Vumc-HD-145-TR930  11-Vumc-HD-149-TR932  ...  \\\n",
       "ENSG00000000419                   18                    28  ...   \n",
       "ENSG00000000460                    5                     7  ...   \n",
       "ENSG00000000938                   19                    36  ...   \n",
       "\n",
       "                 countMatrix.3503-NKI-4707-SARC-TR3079  \\\n",
       "ENSG00000000419                                      0   \n",
       "ENSG00000000460                                      3   \n",
       "ENSG00000000938                                     45   \n",
       "\n",
       "                 countMatrix.3504-NKI-4941-SARC-TR3084  \\\n",
       "ENSG00000000419                                      1   \n",
       "ENSG00000000460                                      0   \n",
       "ENSG00000000938                                      1   \n",
       "\n",
       "                 countMatrix.3507-NKI-6348-SARC-TR3134  \\\n",
       "ENSG00000000419                                      2   \n",
       "ENSG00000000460                                      2   \n",
       "ENSG00000000938                                      7   \n",
       "\n",
       "                 countMatrix.3508-NKI-5275-SARC-TR3099  \\\n",
       "ENSG00000000419                                      9   \n",
       "ENSG00000000460                                     63   \n",
       "ENSG00000000938                                    163   \n",
       "\n",
       "                 countMatrix.3512-NKI-5148-SARC-TR3129  countMatrix.320-10  \\\n",
       "ENSG00000000419                                     16                  22   \n",
       "ENSG00000000460                                      3                   3   \n",
       "ENSG00000000938                                     90                  46   \n",
       "\n",
       "                 countMatrix.327-34  countMatrix.376-P-PH-001-013-TR1400  \\\n",
       "ENSG00000000419                  21                                    0   \n",
       "ENSG00000000460                   0                                    1   \n",
       "ENSG00000000938                  75                                   14   \n",
       "\n",
       "                 countMatrix.506-NKI-NSCLC-107-270  Length  \n",
       "ENSG00000000419                                  1    1075  \n",
       "ENSG00000000460                                  0    4355  \n",
       "ENSG00000000938                                  8    2729  \n",
       "\n",
       "[3 rows x 2352 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add information about length in df_data\n",
    "# Merge: inner (keep only genes that appear in both databases)\n",
    "df_data_wLength = pd.merge(df_data, df_genes_length, left_index = True, right_index = True, how='inner')\n",
    "print(\"Size: \", df_data_wLength.shape)\n",
    "df_data_wLength.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1-Vumc-HD-101-TR922</th>\n",
       "      <th>2-Vumc-HD-103-TR923</th>\n",
       "      <th>3-Vumc-HD-108-TR924</th>\n",
       "      <th>4-Vumc-HD-127-TR925</th>\n",
       "      <th>5-Vumc-HD-130-TR926</th>\n",
       "      <th>6-Vumc-HD-142-TR927</th>\n",
       "      <th>7-Vumc-HD-143-TR928</th>\n",
       "      <th>8-Vumc-HD-144-TR929</th>\n",
       "      <th>9-Vumc-HD-145-TR930</th>\n",
       "      <th>11-Vumc-HD-149-TR932</th>\n",
       "      <th>...</th>\n",
       "      <th>countMatrix.3502-NKI-4854-SARC-TR3080</th>\n",
       "      <th>countMatrix.3503-NKI-4707-SARC-TR3079</th>\n",
       "      <th>countMatrix.3504-NKI-4941-SARC-TR3084</th>\n",
       "      <th>countMatrix.3507-NKI-6348-SARC-TR3134</th>\n",
       "      <th>countMatrix.3508-NKI-5275-SARC-TR3099</th>\n",
       "      <th>countMatrix.3512-NKI-5148-SARC-TR3129</th>\n",
       "      <th>countMatrix.320-10</th>\n",
       "      <th>countMatrix.327-34</th>\n",
       "      <th>countMatrix.376-P-PH-001-013-TR1400</th>\n",
       "      <th>countMatrix.506-NKI-NSCLC-107-270</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>4.651163</td>\n",
       "      <td>73.488372</td>\n",
       "      <td>53.953488</td>\n",
       "      <td>119.069767</td>\n",
       "      <td>86.511628</td>\n",
       "      <td>11.162791</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>16.744186</td>\n",
       "      <td>26.046512</td>\n",
       "      <td>...</td>\n",
       "      <td>2.790698</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.930233</td>\n",
       "      <td>1.860465</td>\n",
       "      <td>8.372093</td>\n",
       "      <td>14.883721</td>\n",
       "      <td>20.465116</td>\n",
       "      <td>19.534884</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.930233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>1.836969</td>\n",
       "      <td>2.296211</td>\n",
       "      <td>4.822044</td>\n",
       "      <td>3.214696</td>\n",
       "      <td>1.377727</td>\n",
       "      <td>1.836969</td>\n",
       "      <td>5.740528</td>\n",
       "      <td>9.644087</td>\n",
       "      <td>1.148106</td>\n",
       "      <td>1.607348</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.688863</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.459242</td>\n",
       "      <td>14.466131</td>\n",
       "      <td>0.688863</td>\n",
       "      <td>0.688863</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.229621</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>8.794430</td>\n",
       "      <td>123.122023</td>\n",
       "      <td>122.389154</td>\n",
       "      <td>21.619641</td>\n",
       "      <td>43.972151</td>\n",
       "      <td>8.061561</td>\n",
       "      <td>11.359472</td>\n",
       "      <td>4.763650</td>\n",
       "      <td>6.962257</td>\n",
       "      <td>13.191645</td>\n",
       "      <td>...</td>\n",
       "      <td>2.931477</td>\n",
       "      <td>16.489557</td>\n",
       "      <td>0.366435</td>\n",
       "      <td>2.565042</td>\n",
       "      <td>59.728838</td>\n",
       "      <td>32.979113</td>\n",
       "      <td>16.855991</td>\n",
       "      <td>27.482594</td>\n",
       "      <td>5.130084</td>\n",
       "      <td>2.931477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <td>16.129032</td>\n",
       "      <td>19.524618</td>\n",
       "      <td>6.366723</td>\n",
       "      <td>15.280136</td>\n",
       "      <td>10.611205</td>\n",
       "      <td>18.251273</td>\n",
       "      <td>11.460102</td>\n",
       "      <td>11.460102</td>\n",
       "      <td>25.042445</td>\n",
       "      <td>2.122241</td>\n",
       "      <td>...</td>\n",
       "      <td>0.424448</td>\n",
       "      <td>1.273345</td>\n",
       "      <td>1.697793</td>\n",
       "      <td>6.366723</td>\n",
       "      <td>1.273345</td>\n",
       "      <td>2.546689</td>\n",
       "      <td>4.244482</td>\n",
       "      <td>3.395586</td>\n",
       "      <td>1.273345</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001461</th>\n",
       "      <td>13.136289</td>\n",
       "      <td>3.101624</td>\n",
       "      <td>2.006933</td>\n",
       "      <td>1.642036</td>\n",
       "      <td>3.284072</td>\n",
       "      <td>5.108557</td>\n",
       "      <td>0.912242</td>\n",
       "      <td>1.094691</td>\n",
       "      <td>8.939974</td>\n",
       "      <td>3.284072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.182448</td>\n",
       "      <td>0.182448</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.189381</td>\n",
       "      <td>0.364897</td>\n",
       "      <td>1.094691</td>\n",
       "      <td>1.094691</td>\n",
       "      <td>0.182448</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2351 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 1-Vumc-HD-101-TR922  2-Vumc-HD-103-TR923  \\\n",
       "ENSG00000000419             4.651163            73.488372   \n",
       "ENSG00000000460             1.836969             2.296211   \n",
       "ENSG00000000938             8.794430           123.122023   \n",
       "ENSG00000001036            16.129032            19.524618   \n",
       "ENSG00000001461            13.136289             3.101624   \n",
       "\n",
       "                 3-Vumc-HD-108-TR924  4-Vumc-HD-127-TR925  \\\n",
       "ENSG00000000419            53.953488           119.069767   \n",
       "ENSG00000000460             4.822044             3.214696   \n",
       "ENSG00000000938           122.389154            21.619641   \n",
       "ENSG00000001036             6.366723            15.280136   \n",
       "ENSG00000001461             2.006933             1.642036   \n",
       "\n",
       "                 5-Vumc-HD-130-TR926  6-Vumc-HD-142-TR927  \\\n",
       "ENSG00000000419            86.511628            11.162791   \n",
       "ENSG00000000460             1.377727             1.836969   \n",
       "ENSG00000000938            43.972151             8.061561   \n",
       "ENSG00000001036            10.611205            18.251273   \n",
       "ENSG00000001461             3.284072             5.108557   \n",
       "\n",
       "                 7-Vumc-HD-143-TR928  8-Vumc-HD-144-TR929  \\\n",
       "ENSG00000000419             0.000000             0.000000   \n",
       "ENSG00000000460             5.740528             9.644087   \n",
       "ENSG00000000938            11.359472             4.763650   \n",
       "ENSG00000001036            11.460102            11.460102   \n",
       "ENSG00000001461             0.912242             1.094691   \n",
       "\n",
       "                 9-Vumc-HD-145-TR930  11-Vumc-HD-149-TR932  ...  \\\n",
       "ENSG00000000419            16.744186             26.046512  ...   \n",
       "ENSG00000000460             1.148106              1.607348  ...   \n",
       "ENSG00000000938             6.962257             13.191645  ...   \n",
       "ENSG00000001036            25.042445              2.122241  ...   \n",
       "ENSG00000001461             8.939974              3.284072  ...   \n",
       "\n",
       "                 countMatrix.3502-NKI-4854-SARC-TR3080  \\\n",
       "ENSG00000000419                               2.790698   \n",
       "ENSG00000000460                               0.000000   \n",
       "ENSG00000000938                               2.931477   \n",
       "ENSG00000001036                               0.424448   \n",
       "ENSG00000001461                               0.182448   \n",
       "\n",
       "                 countMatrix.3503-NKI-4707-SARC-TR3079  \\\n",
       "ENSG00000000419                               0.000000   \n",
       "ENSG00000000460                               0.688863   \n",
       "ENSG00000000938                              16.489557   \n",
       "ENSG00000001036                               1.273345   \n",
       "ENSG00000001461                               0.182448   \n",
       "\n",
       "                 countMatrix.3504-NKI-4941-SARC-TR3084  \\\n",
       "ENSG00000000419                               0.930233   \n",
       "ENSG00000000460                               0.000000   \n",
       "ENSG00000000938                               0.366435   \n",
       "ENSG00000001036                               1.697793   \n",
       "ENSG00000001461                               0.000000   \n",
       "\n",
       "                 countMatrix.3507-NKI-6348-SARC-TR3134  \\\n",
       "ENSG00000000419                               1.860465   \n",
       "ENSG00000000460                               0.459242   \n",
       "ENSG00000000938                               2.565042   \n",
       "ENSG00000001036                               6.366723   \n",
       "ENSG00000001461                               0.000000   \n",
       "\n",
       "                 countMatrix.3508-NKI-5275-SARC-TR3099  \\\n",
       "ENSG00000000419                               8.372093   \n",
       "ENSG00000000460                              14.466131   \n",
       "ENSG00000000938                              59.728838   \n",
       "ENSG00000001036                               1.273345   \n",
       "ENSG00000001461                               2.189381   \n",
       "\n",
       "                 countMatrix.3512-NKI-5148-SARC-TR3129  countMatrix.320-10  \\\n",
       "ENSG00000000419                              14.883721           20.465116   \n",
       "ENSG00000000460                               0.688863            0.688863   \n",
       "ENSG00000000938                              32.979113           16.855991   \n",
       "ENSG00000001036                               2.546689            4.244482   \n",
       "ENSG00000001461                               0.364897            1.094691   \n",
       "\n",
       "                 countMatrix.327-34  countMatrix.376-P-PH-001-013-TR1400  \\\n",
       "ENSG00000000419           19.534884                             0.000000   \n",
       "ENSG00000000460            0.000000                             0.229621   \n",
       "ENSG00000000938           27.482594                             5.130084   \n",
       "ENSG00000001036            3.395586                             1.273345   \n",
       "ENSG00000001461            1.094691                             0.182448   \n",
       "\n",
       "                 countMatrix.506-NKI-NSCLC-107-270  \n",
       "ENSG00000000419                           0.930233  \n",
       "ENSG00000000460                           0.000000  \n",
       "ENSG00000000938                           2.931477  \n",
       "ENSG00000001036                           0.000000  \n",
       "ENSG00000001461                           0.000000  \n",
       "\n",
       "[5 rows x 2351 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apply TPM\n",
    "# Step 1: Normalize for gene length (Kb)\n",
    "# For each row, divide each element by transcript length\n",
    "col_names = list(df_data_wLength.columns)\n",
    "df_data_RFK = 1000*df_data_wLength[col_names[0:-1]].div(df_data_wLength[col_names[-1]], axis = 0)\n",
    "\n",
    "df_data_RFK.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1-Vumc-HD-101-TR922</th>\n",
       "      <th>2-Vumc-HD-103-TR923</th>\n",
       "      <th>3-Vumc-HD-108-TR924</th>\n",
       "      <th>4-Vumc-HD-127-TR925</th>\n",
       "      <th>5-Vumc-HD-130-TR926</th>\n",
       "      <th>6-Vumc-HD-142-TR927</th>\n",
       "      <th>7-Vumc-HD-143-TR928</th>\n",
       "      <th>8-Vumc-HD-144-TR929</th>\n",
       "      <th>9-Vumc-HD-145-TR930</th>\n",
       "      <th>11-Vumc-HD-149-TR932</th>\n",
       "      <th>...</th>\n",
       "      <th>countMatrix.3502-NKI-4854-SARC-TR3080</th>\n",
       "      <th>countMatrix.3503-NKI-4707-SARC-TR3079</th>\n",
       "      <th>countMatrix.3504-NKI-4941-SARC-TR3084</th>\n",
       "      <th>countMatrix.3507-NKI-6348-SARC-TR3134</th>\n",
       "      <th>countMatrix.3508-NKI-5275-SARC-TR3099</th>\n",
       "      <th>countMatrix.3512-NKI-5148-SARC-TR3129</th>\n",
       "      <th>countMatrix.320-10</th>\n",
       "      <th>countMatrix.327-34</th>\n",
       "      <th>countMatrix.376-P-PH-001-013-TR1400</th>\n",
       "      <th>countMatrix.506-NKI-NSCLC-107-270</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>3.937800</td>\n",
       "      <td>67.841060</td>\n",
       "      <td>82.445588</td>\n",
       "      <td>144.198227</td>\n",
       "      <td>85.057059</td>\n",
       "      <td>9.972581</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.591351</td>\n",
       "      <td>31.777072</td>\n",
       "      <td>...</td>\n",
       "      <td>100.611799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>27.876966</td>\n",
       "      <td>7.352040</td>\n",
       "      <td>42.557469</td>\n",
       "      <td>102.610884</td>\n",
       "      <td>113.742517</td>\n",
       "      <td>109.494730</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.438796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>1.555228</td>\n",
       "      <td>2.119756</td>\n",
       "      <td>7.368499</td>\n",
       "      <td>3.893125</td>\n",
       "      <td>1.354562</td>\n",
       "      <td>1.641106</td>\n",
       "      <td>4.863694</td>\n",
       "      <td>10.302517</td>\n",
       "      <td>0.863356</td>\n",
       "      <td>1.960985</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.382223</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.814797</td>\n",
       "      <td>73.535006</td>\n",
       "      <td>4.749140</td>\n",
       "      <td>3.828615</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.622014</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>7.445603</td>\n",
       "      <td>113.660546</td>\n",
       "      <td>187.021192</td>\n",
       "      <td>26.182246</td>\n",
       "      <td>43.232822</td>\n",
       "      <td>7.202014</td>\n",
       "      <td>9.624375</td>\n",
       "      <td>5.088878</td>\n",
       "      <td>5.235502</td>\n",
       "      <td>16.093973</td>\n",
       "      <td>...</td>\n",
       "      <td>105.687245</td>\n",
       "      <td>248.522789</td>\n",
       "      <td>10.981216</td>\n",
       "      <td>10.136332</td>\n",
       "      <td>303.616808</td>\n",
       "      <td>227.363572</td>\n",
       "      <td>93.683458</td>\n",
       "      <td>154.042343</td>\n",
       "      <td>58.579768</td>\n",
       "      <td>64.409543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <td>13.655275</td>\n",
       "      <td>18.024223</td>\n",
       "      <td>9.728903</td>\n",
       "      <td>18.504853</td>\n",
       "      <td>10.432793</td>\n",
       "      <td>16.305269</td>\n",
       "      <td>9.709634</td>\n",
       "      <td>12.242516</td>\n",
       "      <td>18.831504</td>\n",
       "      <td>2.589161</td>\n",
       "      <td>...</td>\n",
       "      <td>15.302445</td>\n",
       "      <td>19.191248</td>\n",
       "      <td>50.879012</td>\n",
       "      <td>25.159517</td>\n",
       "      <td>6.472733</td>\n",
       "      <td>17.557306</td>\n",
       "      <td>23.590293</td>\n",
       "      <td>19.032555</td>\n",
       "      <td>14.540158</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001461</th>\n",
       "      <td>11.121538</td>\n",
       "      <td>2.863275</td>\n",
       "      <td>3.066767</td>\n",
       "      <td>1.988571</td>\n",
       "      <td>3.228855</td>\n",
       "      <td>4.563867</td>\n",
       "      <td>0.772902</td>\n",
       "      <td>1.169428</td>\n",
       "      <td>6.722713</td>\n",
       "      <td>4.006609</td>\n",
       "      <td>...</td>\n",
       "      <td>6.577734</td>\n",
       "      <td>2.749777</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.129180</td>\n",
       "      <td>2.515661</td>\n",
       "      <td>6.084152</td>\n",
       "      <td>6.135837</td>\n",
       "      <td>2.083355</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2351 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 1-Vumc-HD-101-TR922  2-Vumc-HD-103-TR923  \\\n",
       "ENSG00000000419             3.937800            67.841060   \n",
       "ENSG00000000460             1.555228             2.119756   \n",
       "ENSG00000000938             7.445603           113.660546   \n",
       "ENSG00000001036            13.655275            18.024223   \n",
       "ENSG00000001461            11.121538             2.863275   \n",
       "\n",
       "                 3-Vumc-HD-108-TR924  4-Vumc-HD-127-TR925  \\\n",
       "ENSG00000000419            82.445588           144.198227   \n",
       "ENSG00000000460             7.368499             3.893125   \n",
       "ENSG00000000938           187.021192            26.182246   \n",
       "ENSG00000001036             9.728903            18.504853   \n",
       "ENSG00000001461             3.066767             1.988571   \n",
       "\n",
       "                 5-Vumc-HD-130-TR926  6-Vumc-HD-142-TR927  \\\n",
       "ENSG00000000419            85.057059             9.972581   \n",
       "ENSG00000000460             1.354562             1.641106   \n",
       "ENSG00000000938            43.232822             7.202014   \n",
       "ENSG00000001036            10.432793            16.305269   \n",
       "ENSG00000001461             3.228855             4.563867   \n",
       "\n",
       "                 7-Vumc-HD-143-TR928  8-Vumc-HD-144-TR929  \\\n",
       "ENSG00000000419             0.000000             0.000000   \n",
       "ENSG00000000460             4.863694            10.302517   \n",
       "ENSG00000000938             9.624375             5.088878   \n",
       "ENSG00000001036             9.709634            12.242516   \n",
       "ENSG00000001461             0.772902             1.169428   \n",
       "\n",
       "                 9-Vumc-HD-145-TR930  11-Vumc-HD-149-TR932  ...  \\\n",
       "ENSG00000000419            12.591351             31.777072  ...   \n",
       "ENSG00000000460             0.863356              1.960985  ...   \n",
       "ENSG00000000938             5.235502             16.093973  ...   \n",
       "ENSG00000001036            18.831504              2.589161  ...   \n",
       "ENSG00000001461             6.722713              4.006609  ...   \n",
       "\n",
       "                 countMatrix.3502-NKI-4854-SARC-TR3080  \\\n",
       "ENSG00000000419                             100.611799   \n",
       "ENSG00000000460                               0.000000   \n",
       "ENSG00000000938                             105.687245   \n",
       "ENSG00000001036                              15.302445   \n",
       "ENSG00000001461                               6.577734   \n",
       "\n",
       "                 countMatrix.3503-NKI-4707-SARC-TR3079  \\\n",
       "ENSG00000000419                               0.000000   \n",
       "ENSG00000000460                              10.382223   \n",
       "ENSG00000000938                             248.522789   \n",
       "ENSG00000001036                              19.191248   \n",
       "ENSG00000001461                               2.749777   \n",
       "\n",
       "                 countMatrix.3504-NKI-4941-SARC-TR3084  \\\n",
       "ENSG00000000419                              27.876966   \n",
       "ENSG00000000460                               0.000000   \n",
       "ENSG00000000938                              10.981216   \n",
       "ENSG00000001036                              50.879012   \n",
       "ENSG00000001461                               0.000000   \n",
       "\n",
       "                 countMatrix.3507-NKI-6348-SARC-TR3134  \\\n",
       "ENSG00000000419                               7.352040   \n",
       "ENSG00000000460                               1.814797   \n",
       "ENSG00000000938                              10.136332   \n",
       "ENSG00000001036                              25.159517   \n",
       "ENSG00000001461                               0.000000   \n",
       "\n",
       "                 countMatrix.3508-NKI-5275-SARC-TR3099  \\\n",
       "ENSG00000000419                              42.557469   \n",
       "ENSG00000000460                              73.535006   \n",
       "ENSG00000000938                             303.616808   \n",
       "ENSG00000001036                               6.472733   \n",
       "ENSG00000001461                              11.129180   \n",
       "\n",
       "                 countMatrix.3512-NKI-5148-SARC-TR3129  countMatrix.320-10  \\\n",
       "ENSG00000000419                             102.610884          113.742517   \n",
       "ENSG00000000460                               4.749140            3.828615   \n",
       "ENSG00000000938                             227.363572           93.683458   \n",
       "ENSG00000001036                              17.557306           23.590293   \n",
       "ENSG00000001461                               2.515661            6.084152   \n",
       "\n",
       "                 countMatrix.327-34  countMatrix.376-P-PH-001-013-TR1400  \\\n",
       "ENSG00000000419          109.494730                             0.000000   \n",
       "ENSG00000000460            0.000000                             2.622014   \n",
       "ENSG00000000938          154.042343                            58.579768   \n",
       "ENSG00000001036           19.032555                            14.540158   \n",
       "ENSG00000001461            6.135837                             2.083355   \n",
       "\n",
       "                 countMatrix.506-NKI-NSCLC-107-270  \n",
       "ENSG00000000419                          20.438796  \n",
       "ENSG00000000460                           0.000000  \n",
       "ENSG00000000938                          64.409543  \n",
       "ENSG00000001036                           0.000000  \n",
       "ENSG00000001461                           0.000000  \n",
       "\n",
       "[5 rows x 2351 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Step 2: Normalize for sequencing depth (M)\n",
    "# For each column: \n",
    "#    sum all entries and divide by 10^6\n",
    "#    then, divide each element by the above number\n",
    "\n",
    "# Total = data_df_present_wLength.sum()\n",
    "# Total = Total / 10**6\n",
    "\n",
    "df_data_TPM = df_data_RFK.transform(lambda x: (x*10**6)/x.sum())\n",
    "\n",
    "df_data_TPM.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5251, 2351)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data_TPM.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1-Vumc-HD-101-TR922                      1000000.0\n",
       "2-Vumc-HD-103-TR923                      1000000.0\n",
       "3-Vumc-HD-108-TR924                      1000000.0\n",
       "4-Vumc-HD-127-TR925                      1000000.0\n",
       "5-Vumc-HD-130-TR926                      1000000.0\n",
       "                                           ...    \n",
       "countMatrix.3512-NKI-5148-SARC-TR3129    1000000.0\n",
       "countMatrix.320-10                       1000000.0\n",
       "countMatrix.327-34                       1000000.0\n",
       "countMatrix.376-P-PH-001-013-TR1400      1000000.0\n",
       "countMatrix.506-NKI-NSCLC-107-270        1000000.0\n",
       "Length: 2351, dtype: float64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Verify TPM is correctly done by checking each patient (column) adds up 1M\n",
    "df_data_TPM.sum()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_data = df_data_TPM.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001461</th>\n",
       "      <th>ENSG00000001497</th>\n",
       "      <th>ENSG00000001629</th>\n",
       "      <th>ENSG00000001631</th>\n",
       "      <th>ENSG00000002330</th>\n",
       "      <th>ENSG00000002549</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000257267</th>\n",
       "      <th>ENSG00000257923</th>\n",
       "      <th>ENSG00000258890</th>\n",
       "      <th>ENSG00000261253</th>\n",
       "      <th>ENSG00000263563</th>\n",
       "      <th>ENSG00000264538</th>\n",
       "      <th>ENSG00000265148</th>\n",
       "      <th>ENSG00000266714</th>\n",
       "      <th>ENSG00000267243</th>\n",
       "      <th>ENSG00000271043</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1-Vumc-HD-101-TR922</th>\n",
       "      <td>3.937800</td>\n",
       "      <td>1.555228</td>\n",
       "      <td>7.445603</td>\n",
       "      <td>13.655275</td>\n",
       "      <td>11.121538</td>\n",
       "      <td>3.471206</td>\n",
       "      <td>5.986674</td>\n",
       "      <td>4.978068</td>\n",
       "      <td>46.477912</td>\n",
       "      <td>25.632179</td>\n",
       "      <td>...</td>\n",
       "      <td>69.065205</td>\n",
       "      <td>66.636392</td>\n",
       "      <td>4.045685</td>\n",
       "      <td>180.566933</td>\n",
       "      <td>26.692834</td>\n",
       "      <td>24.964644</td>\n",
       "      <td>53.934968</td>\n",
       "      <td>0.539998</td>\n",
       "      <td>32.907766</td>\n",
       "      <td>284.930419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-Vumc-HD-103-TR923</th>\n",
       "      <td>67.841060</td>\n",
       "      <td>2.119756</td>\n",
       "      <td>113.660546</td>\n",
       "      <td>18.024223</td>\n",
       "      <td>2.863275</td>\n",
       "      <td>15.139871</td>\n",
       "      <td>3.036190</td>\n",
       "      <td>18.222689</td>\n",
       "      <td>40.741981</td>\n",
       "      <td>42.549333</td>\n",
       "      <td>...</td>\n",
       "      <td>30.267333</td>\n",
       "      <td>14.505070</td>\n",
       "      <td>26.468248</td>\n",
       "      <td>59.756005</td>\n",
       "      <td>23.411025</td>\n",
       "      <td>10.651773</td>\n",
       "      <td>32.718322</td>\n",
       "      <td>11.972440</td>\n",
       "      <td>16.903231</td>\n",
       "      <td>118.398686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-Vumc-HD-108-TR924</th>\n",
       "      <td>82.445588</td>\n",
       "      <td>7.368499</td>\n",
       "      <td>187.021192</td>\n",
       "      <td>9.728903</td>\n",
       "      <td>3.066767</td>\n",
       "      <td>7.518260</td>\n",
       "      <td>23.872422</td>\n",
       "      <td>11.552102</td>\n",
       "      <td>24.673084</td>\n",
       "      <td>66.978931</td>\n",
       "      <td>...</td>\n",
       "      <td>24.454153</td>\n",
       "      <td>28.678713</td>\n",
       "      <td>18.498656</td>\n",
       "      <td>97.645350</td>\n",
       "      <td>25.136444</td>\n",
       "      <td>9.795425</td>\n",
       "      <td>61.013765</td>\n",
       "      <td>24.528653</td>\n",
       "      <td>6.381343</td>\n",
       "      <td>145.882880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4-Vumc-HD-127-TR925</th>\n",
       "      <td>144.198227</td>\n",
       "      <td>3.893125</td>\n",
       "      <td>26.182246</td>\n",
       "      <td>18.504853</td>\n",
       "      <td>1.988571</td>\n",
       "      <td>6.454907</td>\n",
       "      <td>3.385574</td>\n",
       "      <td>10.172531</td>\n",
       "      <td>33.893471</td>\n",
       "      <td>22.436797</td>\n",
       "      <td>...</td>\n",
       "      <td>21.507537</td>\n",
       "      <td>5.021406</td>\n",
       "      <td>18.518608</td>\n",
       "      <td>25.125307</td>\n",
       "      <td>19.091100</td>\n",
       "      <td>19.407689</td>\n",
       "      <td>10.866216</td>\n",
       "      <td>18.667032</td>\n",
       "      <td>1.556106</td>\n",
       "      <td>101.601218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5-Vumc-HD-130-TR926</th>\n",
       "      <td>85.057059</td>\n",
       "      <td>1.354562</td>\n",
       "      <td>43.232822</td>\n",
       "      <td>10.432793</td>\n",
       "      <td>3.228855</td>\n",
       "      <td>3.627994</td>\n",
       "      <td>2.748589</td>\n",
       "      <td>10.323251</td>\n",
       "      <td>39.158125</td>\n",
       "      <td>47.981986</td>\n",
       "      <td>...</td>\n",
       "      <td>42.213313</td>\n",
       "      <td>6.222246</td>\n",
       "      <td>20.672286</td>\n",
       "      <td>25.293593</td>\n",
       "      <td>20.890185</td>\n",
       "      <td>8.193220</td>\n",
       "      <td>19.407897</td>\n",
       "      <td>1.254198</td>\n",
       "      <td>0.947497</td>\n",
       "      <td>162.126211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 5251 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ENSG00000000419  ENSG00000000460  ENSG00000000938  \\\n",
       "1-Vumc-HD-101-TR922         3.937800         1.555228         7.445603   \n",
       "2-Vumc-HD-103-TR923        67.841060         2.119756       113.660546   \n",
       "3-Vumc-HD-108-TR924        82.445588         7.368499       187.021192   \n",
       "4-Vumc-HD-127-TR925       144.198227         3.893125        26.182246   \n",
       "5-Vumc-HD-130-TR926        85.057059         1.354562        43.232822   \n",
       "\n",
       "                     ENSG00000001036  ENSG00000001461  ENSG00000001497  \\\n",
       "1-Vumc-HD-101-TR922        13.655275        11.121538         3.471206   \n",
       "2-Vumc-HD-103-TR923        18.024223         2.863275        15.139871   \n",
       "3-Vumc-HD-108-TR924         9.728903         3.066767         7.518260   \n",
       "4-Vumc-HD-127-TR925        18.504853         1.988571         6.454907   \n",
       "5-Vumc-HD-130-TR926        10.432793         3.228855         3.627994   \n",
       "\n",
       "                     ENSG00000001629  ENSG00000001631  ENSG00000002330  \\\n",
       "1-Vumc-HD-101-TR922         5.986674         4.978068        46.477912   \n",
       "2-Vumc-HD-103-TR923         3.036190        18.222689        40.741981   \n",
       "3-Vumc-HD-108-TR924        23.872422        11.552102        24.673084   \n",
       "4-Vumc-HD-127-TR925         3.385574        10.172531        33.893471   \n",
       "5-Vumc-HD-130-TR926         2.748589        10.323251        39.158125   \n",
       "\n",
       "                     ENSG00000002549  ...  ENSG00000257267  ENSG00000257923  \\\n",
       "1-Vumc-HD-101-TR922        25.632179  ...        69.065205        66.636392   \n",
       "2-Vumc-HD-103-TR923        42.549333  ...        30.267333        14.505070   \n",
       "3-Vumc-HD-108-TR924        66.978931  ...        24.454153        28.678713   \n",
       "4-Vumc-HD-127-TR925        22.436797  ...        21.507537         5.021406   \n",
       "5-Vumc-HD-130-TR926        47.981986  ...        42.213313         6.222246   \n",
       "\n",
       "                     ENSG00000258890  ENSG00000261253  ENSG00000263563  \\\n",
       "1-Vumc-HD-101-TR922         4.045685       180.566933        26.692834   \n",
       "2-Vumc-HD-103-TR923        26.468248        59.756005        23.411025   \n",
       "3-Vumc-HD-108-TR924        18.498656        97.645350        25.136444   \n",
       "4-Vumc-HD-127-TR925        18.518608        25.125307        19.091100   \n",
       "5-Vumc-HD-130-TR926        20.672286        25.293593        20.890185   \n",
       "\n",
       "                     ENSG00000264538  ENSG00000265148  ENSG00000266714  \\\n",
       "1-Vumc-HD-101-TR922        24.964644        53.934968         0.539998   \n",
       "2-Vumc-HD-103-TR923        10.651773        32.718322        11.972440   \n",
       "3-Vumc-HD-108-TR924         9.795425        61.013765        24.528653   \n",
       "4-Vumc-HD-127-TR925        19.407689        10.866216        18.667032   \n",
       "5-Vumc-HD-130-TR926         8.193220        19.407897         1.254198   \n",
       "\n",
       "                     ENSG00000267243  ENSG00000271043  \n",
       "1-Vumc-HD-101-TR922        32.907766       284.930419  \n",
       "2-Vumc-HD-103-TR923        16.903231       118.398686  \n",
       "3-Vumc-HD-108-TR924         6.381343       145.882880  \n",
       "4-Vumc-HD-127-TR925         1.556106       101.601218  \n",
       "5-Vumc-HD-130-TR926         0.947497       162.126211  \n",
       "\n",
       "[5 rows x 5251 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(df_data.head(5))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Look for Nan Values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2351, 5251)\n",
      "Genes:  Index(['ENSG00000000419', 'ENSG00000000460', 'ENSG00000000938',\n",
      "       'ENSG00000001036', 'ENSG00000001461', 'ENSG00000001497',\n",
      "       'ENSG00000001629', 'ENSG00000001631', 'ENSG00000002330',\n",
      "       'ENSG00000002549'],\n",
      "      dtype='object')\n",
      "Pacients:  Index(['1-Vumc-HD-101-TR922', '2-Vumc-HD-103-TR923', '3-Vumc-HD-108-TR924',\n",
      "       '4-Vumc-HD-127-TR925', '5-Vumc-HD-130-TR926', '6-Vumc-HD-142-TR927',\n",
      "       '7-Vumc-HD-143-TR928', '8-Vumc-HD-144-TR929', '9-Vumc-HD-145-TR930',\n",
      "       '11-Vumc-HD-149-TR932'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(df_data.shape)\n",
    "genes = df_data.columns\n",
    "pacients = df_data.index\n",
    "print(\"Genes: \", genes[:10])\n",
    "print(\"Pacients: \", pacients[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Verify data type, Naans etc\n",
    "# print(df_data.info())\n",
    "# print(df_data.describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data.dtypes\n",
    "df_data.isnull().values.any()\n",
    "# df_data.isnull().sum()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create DataFrame from metadata (ground truth)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load txt file\n",
    "file = 'IntVeld2022_metadata.txt'\n",
    "\n",
    "with open (file, \"r\") as f:\n",
    "    y_data = f.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2351\n",
      "['NKI-NSCLC-4861-TR2958', 'Malignant', 'Non-Small-Cell Lung Cancer']\n"
     ]
    }
   ],
   "source": [
    "# Change names to match those in the excel file\n",
    "ys_data = []\n",
    "line = 0\n",
    "\n",
    "Group_txt = ['Asymptomatic Controls', 'Pancreatic Disease', 'Urothelial Carcinoma', 'Epilepsy', 'Esophageal Cancer', 'Renal Cell Carcinoma', 'Hodgkin Lymphoma']\n",
    "Group_xml = ['Asymptomatic controls', 'Pancreatic diseases', 'Urothelial cancer', 'Medically-intractable epilepsy', 'Esophageal carcinoma', 'Renal Cell cancer', 'Lymphoma']\n",
    "\n",
    "\n",
    "while line < len(y_data):\n",
    "    if y_data[line]:\n",
    "        if y_data[line][0].isdigit(): \n",
    "            Patient = re.split('[0-9]+: ', y_data[line])[1]\n",
    "            Class_group = re.split('\"',y_data[line+6])[1]\n",
    "            Patient_group = re.split('\"',y_data[line+7])[1]\n",
    "            if Patient_group in Group_txt:\n",
    "                Patient_group = Group_xml[Group_txt.index(Patient_group)]\n",
    "            ys_data.append([Patient,Class_group,Patient_group])\n",
    "            line += 9\n",
    "    line += 1\n",
    "\n",
    "print(len(ys_data))\n",
    "print(ys_data[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Class_group</th>\n",
       "      <th>Patient_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TR3660-OVA-CATH</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Ovarian Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266</th>\n",
       "      <td>TR4253-HN-VUMC-BB</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Head and Neck Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>UMCG-NSCLC-1097-TR2856</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Non-Small-Cell Lung Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1093</th>\n",
       "      <td>Vumc-MS-Presto-021-TR1429</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Multiple Sclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673</th>\n",
       "      <td>Vumc-GBM-follow-up-VU501t0</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Glioma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Sample ID   Class_group               Patient_group\n",
       "19               TR3660-OVA-CATH     Malignant              Ovarian Cancer\n",
       "1266           TR4253-HN-VUMC-BB     Malignant        Head and Neck Cancer\n",
       "193       UMCG-NSCLC-1097-TR2856     Malignant  Non-Small-Cell Lung Cancer\n",
       "1093   Vumc-MS-Presto-021-TR1429  nonMalignant          Multiple Sclerosis\n",
       "673   Vumc-GBM-follow-up-VU501t0     Malignant                      Glioma"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_y = pd.DataFrame(ys_data, columns=['Sample ID', 'Class_group','Patient_group'])\n",
    "df_y.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2351, 3)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_y.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Class_group_values:  ['Malignant' 'nonMalignant']\n",
      "\n",
      "Patient_group_values_Malignant:  ['Asymptomatic controls', 'Breast Cancer', 'Cholangiocarcinoma', 'Colorectal Cancer', 'Endometrial Cancer', 'Esophageal carcinoma', 'Glioma', 'Head and Neck Cancer', 'Hepatocellular Carcinoma', 'Lymphoma', 'Melanoma', 'Multiple Myeloma', 'Non-Small-Cell Lung Cancer', 'Ovarian Cancer', 'Pancreatic Cancer', 'Prostate Cancer', 'Renal Cell cancer', 'Sarcoma', 'Urothelial cancer']\n",
      "Total Patient_group_values_Malignant:  19\n",
      "Number of Malignant samples:  1648\n",
      "\n",
      "Patient_group_values_nonMalignant:  ['Angina Pectoris', 'Asymptomatic controls', 'Bowel Disease', 'Former Sarcoma', 'Hematuria', 'Medically-intractable epilepsy', 'Multiple Sclerosis', 'Pancreatic diseases', 'Prostate Cancer', 'Pulmonary Hypertension', 'nSTEMI']\n",
      "Total Patient_group_values_nonMalignant:  11\n",
      "Number of non Malignant samples:  703\n"
     ]
    }
   ],
   "source": [
    "# Check values per class\n",
    "\n",
    "Class_group_values = df_y['Class_group'].unique()\n",
    "print(\"Class_group_values: \", Class_group_values)\n",
    "\n",
    "Patient_group_values_Malignant = df_y.loc[df_y['Class_group'] == 'Malignant','Patient_group'].unique()\n",
    "print(\"\\nPatient_group_values_Malignant: \", sorted(list(Patient_group_values_Malignant)))\n",
    "print(\"Total Patient_group_values_Malignant: \", len(Patient_group_values_Malignant))\n",
    "print(\"Number of Malignant samples: \" , len(df_y.loc[df_y['Class_group'] == 'Malignant']))\n",
    "\n",
    "Patient_group_values_nonMalignant = df_y.loc[df_y['Class_group'] == 'nonMalignant','Patient_group'].unique()\n",
    "print(\"\\nPatient_group_values_nonMalignant: \", sorted(list(Patient_group_values_nonMalignant)))\n",
    "print(\"Total Patient_group_values_nonMalignant: \", len(Patient_group_values_nonMalignant))\n",
    "print(\"Number of non Malignant samples: \" , len(df_y.loc[df_y['Class_group'] == 'nonMalignant']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples Malignant + Asymtomatic controls:  21\n"
     ]
    }
   ],
   "source": [
    "# Find incorrect data that does not match info from the article\n",
    "df_filtered_y = df_y[(df_y['Class_group']=='Malignant') & (df_y['Patient_group']=='Asymptomatic controls') ]\n",
    "print(\"Number of samples Malignant + Asymtomatic controls: \", len(df_filtered_y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples nonMalignant + Prostate Cancer:  1\n"
     ]
    }
   ],
   "source": [
    "df_filtered_y = df_y[(df_y['Class_group']=='nonMalignant') & (df_y['Patient_group']=='Prostate Cancer') ]\n",
    "print(\"Number of samples nonMalignant + Prostate Cancer: \", len(df_filtered_y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct incorrect data\n",
    "df_y.loc[(df_y['Class_group']=='nonMalignant') & (df_y['Patient_group'] == 'Prostate Cancer'), 'Class_group'] = 'Malignant'\n",
    "df_y.loc[(df_y['Class_group']=='Malignant') & (df_y['Patient_group'] == 'Asymptomatic controls'), 'Class_group'] = 'nonMalignant'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Malignant samples:  1628\n",
      "Number of non Malignant samples:  723\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of Malignant samples: \" , len(df_y.loc[df_y['Class_group'] == 'Malignant']))\n",
    "print(\"Number of non Malignant samples: \" , len(df_y.loc[df_y['Class_group'] == 'nonMalignant']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                            Sample ID  Class_group\n",
      "Patient_group                                     \n",
      "Breast Cancer                      93           93\n",
      "Cholangiocarcinoma                 85           85\n",
      "Colorectal Cancer                  85           85\n",
      "Endometrial Cancer                 39           39\n",
      "Esophageal carcinoma               15           15\n",
      "Glioma                            132          132\n",
      "Head and Neck Cancer              101          101\n",
      "Hepatocellular Carcinoma           23           23\n",
      "Lymphoma                           20           20\n",
      "Melanoma                           68           68\n",
      "Multiple Myeloma                   31           31\n",
      "Non-Small-Cell Lung Cancer        522          522\n",
      "Ovarian Cancer                    144          144\n",
      "Pancreatic Cancer                 126          126\n",
      "Prostate Cancer                    35           35\n",
      "Renal Cell cancer                  28           28\n",
      "Sarcoma                            53           53\n",
      "Urothelial cancer                  28           28\n"
     ]
    }
   ],
   "source": [
    "print(df_y.loc[df_y['Class_group'] == 'Malignant'].groupby(by='Patient_group').count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                Sample ID  Class_group\n",
      "Patient_group                                         \n",
      "Angina Pectoris                        26           26\n",
      "Asymptomatic controls                 391          391\n",
      "Bowel Disease                           9            9\n",
      "Former Sarcoma                         26           26\n",
      "Hematuria                               5            5\n",
      "Medically-intractable epilepsy         44           44\n",
      "Multiple Sclerosis                     86           86\n",
      "Pancreatic diseases                    50           50\n",
      "Pulmonary Hypertension                 71           71\n",
      "nSTEMI                                 15           15\n"
     ]
    }
   ],
   "source": [
    "print(df_y.loc[df_y['Class_group'] == 'nonMalignant'].groupby(by='Patient_group').count())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### XLSX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add information about age, sex, cancer stage and supplying institutions\n",
    "# Compare data between xlsx and txt and correct some information\n",
    "\n",
    "file = '1-s2.0-S1535610822003701-mmc3.xlsx'\n",
    "df_y_xlsx = pd.read_excel(file, header=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample-supplying institution</th>\n",
       "      <th>Training series</th>\n",
       "      <th>Evaluation series</th>\n",
       "      <th>Validation series</th>\n",
       "      <th>lib.size</th>\n",
       "      <th>classificationScoreCancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vumc-HD-101-TR922</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>F</td>\n",
       "      <td>35</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2296824</td>\n",
       "      <td>0.485949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vumc-HD-103-TR923</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>F</td>\n",
       "      <td>54</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1731594</td>\n",
       "      <td>0.370735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Vumc-HD-108-TR924</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>F</td>\n",
       "      <td>56</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1136912</td>\n",
       "      <td>0.799796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vumc-HD-127-TR925</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>F</td>\n",
       "      <td>52</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1115322</td>\n",
       "      <td>0.580309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vumc-HD-130-TR926</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>F</td>\n",
       "      <td>57</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1331137</td>\n",
       "      <td>0.402155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample ID                  Group Stage Sex Age  \\\n",
       "0  Vumc-HD-101-TR922  Asymptomatic controls  n.a.   F  35   \n",
       "1  Vumc-HD-103-TR923  Asymptomatic controls  n.a.   F  54   \n",
       "2  Vumc-HD-108-TR924  Asymptomatic controls  n.a.   F  56   \n",
       "3  Vumc-HD-127-TR925  Asymptomatic controls  n.a.   F  52   \n",
       "4  Vumc-HD-130-TR926  Asymptomatic controls  n.a.   F  57   \n",
       "\n",
       "  Sample-supplying institution  Training series  Evaluation series  \\\n",
       "0                 Institute 13                1                  0   \n",
       "1                 Institute 13                0                  1   \n",
       "2                 Institute 13                0                  0   \n",
       "3                 Institute 13                1                  0   \n",
       "4                 Institute 13                0                  1   \n",
       "\n",
       "   Validation series  lib.size  classificationScoreCancer  \n",
       "0                  0   2296824                   0.485949  \n",
       "1                  0   1731594                   0.370735  \n",
       "2                  1   1136912                   0.799796  \n",
       "3                  0   1115322                   0.580309  \n",
       "4                  0   1331137                   0.402155  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_y_xlsx.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_y_xlsx.replace('n.a.', np.nan, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_y_xlsx.isnull().values.any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Stage', 'Sex', 'Age']\n"
     ]
    }
   ],
   "source": [
    "# Check which columns have Nan values\n",
    "list_na_columns = df_y_xlsx.columns[df_y_xlsx.isna().any()].tolist()\n",
    "print(list_na_columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2351, 13)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge info from txt and xmls\n",
    "df_y_merged = pd.merge(df_y_xlsx, df_y, on=['Sample ID'], how='inner')\n",
    "df_y_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample-supplying institution</th>\n",
       "      <th>Training series</th>\n",
       "      <th>Evaluation series</th>\n",
       "      <th>Validation series</th>\n",
       "      <th>lib.size</th>\n",
       "      <th>classificationScoreCancer</th>\n",
       "      <th>Class_group</th>\n",
       "      <th>Patient_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vumc-HD-101-TR922</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2296824</td>\n",
       "      <td>0.485949</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vumc-HD-103-TR923</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>54.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1731594</td>\n",
       "      <td>0.370735</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Vumc-HD-108-TR924</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1136912</td>\n",
       "      <td>0.799796</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vumc-HD-127-TR925</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>52.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1115322</td>\n",
       "      <td>0.580309</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vumc-HD-130-TR926</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>57.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1331137</td>\n",
       "      <td>0.402155</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample ID                  Group Stage Sex   Age  \\\n",
       "0  Vumc-HD-101-TR922  Asymptomatic controls   NaN   F  35.0   \n",
       "1  Vumc-HD-103-TR923  Asymptomatic controls   NaN   F  54.0   \n",
       "2  Vumc-HD-108-TR924  Asymptomatic controls   NaN   F  56.0   \n",
       "3  Vumc-HD-127-TR925  Asymptomatic controls   NaN   F  52.0   \n",
       "4  Vumc-HD-130-TR926  Asymptomatic controls   NaN   F  57.0   \n",
       "\n",
       "  Sample-supplying institution  Training series  Evaluation series  \\\n",
       "0                 Institute 13                1                  0   \n",
       "1                 Institute 13                0                  1   \n",
       "2                 Institute 13                0                  0   \n",
       "3                 Institute 13                1                  0   \n",
       "4                 Institute 13                0                  1   \n",
       "\n",
       "   Validation series  lib.size  classificationScoreCancer   Class_group  \\\n",
       "0                  0   2296824                   0.485949  nonMalignant   \n",
       "1                  0   1731594                   0.370735  nonMalignant   \n",
       "2                  1   1136912                   0.799796  nonMalignant   \n",
       "3                  0   1115322                   0.580309  nonMalignant   \n",
       "4                  0   1331137                   0.402155  nonMalignant   \n",
       "\n",
       "           Patient_group  \n",
       "0  Asymptomatic controls  \n",
       "1  Asymptomatic controls  \n",
       "2  Asymptomatic controls  \n",
       "3  Asymptomatic controls  \n",
       "4  Asymptomatic controls  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_y_merged.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample-supplying institution</th>\n",
       "      <th>Training series</th>\n",
       "      <th>Evaluation series</th>\n",
       "      <th>Validation series</th>\n",
       "      <th>lib.size</th>\n",
       "      <th>classificationScoreCancer</th>\n",
       "      <th>Class_group</th>\n",
       "      <th>Patient_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2318</th>\n",
       "      <td>Vumc-P-PH-001-024-TR1414</td>\n",
       "      <td>Pulmonary Hypertension</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>71.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>464882</td>\n",
       "      <td>0.949583</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Sample ID                   Group Stage Sex   Age  \\\n",
       "2318  Vumc-P-PH-001-024-TR1414  Pulmonary Hypertension   NaN   M  71.0   \n",
       "\n",
       "     Sample-supplying institution  Training series  Evaluation series  \\\n",
       "2318                 Institute 13                0                  0   \n",
       "\n",
       "      Validation series  lib.size  classificationScoreCancer   Class_group  \\\n",
       "2318                  1    464882                   0.949583  nonMalignant   \n",
       "\n",
       "              Patient_group  \n",
       "2318  Asymptomatic controls  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find differences and change values\n",
    "df_y_merged[df_y_merged[\"Group\"].astype(str).str.lower() != df_y_merged[\"Patient_group\"].astype(str).str.lower()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_y_merged.loc[df_y_merged[\"Group\"].astype(str).str.lower() != df_y_merged[\"Patient_group\"].astype(str).str.lower(),'Patient_group'] = 'Pulmonary Hypertension'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Class_group\n",
       "Malignant       1628\n",
       "nonMalignant     723\n",
       "dtype: int64"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Review numbers and compare with article\n",
    "df_y_merged.groupby('Class_group').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patient_group\n",
      "Breast Cancer                  93\n",
      "Cholangiocarcinoma             85\n",
      "Colorectal Cancer              85\n",
      "Endometrial Cancer             39\n",
      "Esophageal carcinoma           15\n",
      "Glioma                        132\n",
      "Head and Neck Cancer          101\n",
      "Hepatocellular Carcinoma       23\n",
      "Lymphoma                       20\n",
      "Melanoma                       68\n",
      "Multiple Myeloma               31\n",
      "Non-Small-Cell Lung Cancer    522\n",
      "Ovarian Cancer                144\n",
      "Pancreatic Cancer             126\n",
      "Prostate Cancer                35\n",
      "Renal Cell cancer              28\n",
      "Sarcoma                        53\n",
      "Urothelial cancer              28\n",
      "dtype: int64\n",
      "Total number of cancers:  18\n"
     ]
    }
   ],
   "source": [
    "types_of_cancers = df_y_merged[df_y_merged['Class_group'] == 'Malignant'].groupby(by='Patient_group').size()\n",
    "print(types_of_cancers)\n",
    "print(\"Total number of cancers: \", len(types_of_cancers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patient_group\n",
      "Angina Pectoris                    26\n",
      "Asymptomatic controls             390\n",
      "Bowel Disease                       9\n",
      "Former Sarcoma                     26\n",
      "Hematuria                           5\n",
      "Medically-intractable epilepsy     44\n",
      "Multiple Sclerosis                 86\n",
      "Pancreatic diseases                50\n",
      "Pulmonary Hypertension             72\n",
      "nSTEMI                             15\n",
      "dtype: int64\n",
      "Total number of categories of nonMalignant patients (apart from Asymptomatic controls):  9\n"
     ]
    }
   ],
   "source": [
    "types_of_nonCancers = df_y_merged[df_y_merged['Class_group'] == 'nonMalignant'].groupby(by='Patient_group').size()\n",
    "print(types_of_nonCancers)\n",
    "print(\"Total number of categories of nonMalignant patients (apart from Asymptomatic controls): \", len(types_of_nonCancers) - 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage\n",
      "I      111\n",
      "II     166\n",
      "III    235\n",
      "IV     860\n",
      "dtype: int64\n",
      "\n",
      " Series([], dtype: int64)\n"
     ]
    }
   ],
   "source": [
    "# Article: For a subset of samples, tumor stages were unknown (n.a.; n = 124) or not informative (n.i.; e.g.,\n",
    "# gliomas and multiple myeloma; n = 132; in total n = 256; 16% of all cancers).\n",
    "stage_of_cancers = df_y_merged[df_y_merged['Class_group'] == 'Malignant'].groupby(by='Stage').size()\n",
    "print(stage_of_cancers)\n",
    "\n",
    "stage_of_cancers_glioma = df_y_merged[df_y_merged['Patient_group'] == 'Glioma'].groupby(by='Stage').size()\n",
    "print(\"\\n\",stage_of_cancers_glioma)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample-supplying institution</th>\n",
       "      <th>Training series</th>\n",
       "      <th>Evaluation series</th>\n",
       "      <th>Validation series</th>\n",
       "      <th>lib.size</th>\n",
       "      <th>classificationScoreCancer</th>\n",
       "      <th>Class_group</th>\n",
       "      <th>Patient_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>Vumc-ChronPan-29-TR1045</td>\n",
       "      <td>Pancreatic diseases</td>\n",
       "      <td>IV</td>\n",
       "      <td>M</td>\n",
       "      <td>58.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1504224</td>\n",
       "      <td>0.946398</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Pancreatic diseases</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Sample ID                Group Stage Sex   Age  \\\n",
       "789  Vumc-ChronPan-29-TR1045  Pancreatic diseases    IV   M  58.0   \n",
       "\n",
       "    Sample-supplying institution  Training series  Evaluation series  \\\n",
       "789                 Institute 13                0                  0   \n",
       "\n",
       "     Validation series  lib.size  classificationScoreCancer   Class_group  \\\n",
       "789                  1   1504224                   0.946398  nonMalignant   \n",
       "\n",
       "           Patient_group  \n",
       "789  Pancreatic diseases  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Correct data: nonMalignant should not have Stage\n",
    "df_y_merged.loc[(~df_y_merged['Stage'].isna()) & (df_y_merged['Class_group']=='nonMalignant')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_y_merged.loc[(~df_y_merged['Stage'].isna()) & (df_y_merged['Class_group']=='nonMalignant'),'Stage'] = np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample-supplying institution</th>\n",
       "      <th>Training series</th>\n",
       "      <th>Evaluation series</th>\n",
       "      <th>Validation series</th>\n",
       "      <th>Class_group</th>\n",
       "      <th>Patient_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2196</th>\n",
       "      <td>TR4547-SARC-NKI</td>\n",
       "      <td>IV</td>\n",
       "      <td>F</td>\n",
       "      <td>78.0</td>\n",
       "      <td>Institute 5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Sarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1503</th>\n",
       "      <td>Vumc-NSCLC-374-TR2178</td>\n",
       "      <td>IV</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Non-Small-Cell Lung Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>MGH-BrCa-P12-TR703</td>\n",
       "      <td>IV</td>\n",
       "      <td>F</td>\n",
       "      <td>43.0</td>\n",
       "      <td>Institute 4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Breast Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1061</th>\n",
       "      <td>TR4808-NSCLC-MAAS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>76.0</td>\n",
       "      <td>Institute 3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Vumc-HD-21-1-TR1161</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>39.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample ID Stage Sex   Age Sample-supplying institution  \\\n",
       "2196        TR4547-SARC-NKI    IV   F  78.0                  Institute 5   \n",
       "1503  Vumc-NSCLC-374-TR2178    IV   F  56.0                 Institute 13   \n",
       "689      MGH-BrCa-P12-TR703    IV   F  43.0                  Institute 4   \n",
       "1061      TR4808-NSCLC-MAAS   NaN   M  76.0                  Institute 3   \n",
       "86      Vumc-HD-21-1-TR1161   NaN   F  39.0                 Institute 13   \n",
       "\n",
       "      Training series  Evaluation series  Validation series   Class_group  \\\n",
       "2196                0                  1                  0     Malignant   \n",
       "1503                0                  0                  1     Malignant   \n",
       "689                 0                  0                  1     Malignant   \n",
       "1061                0                  0                  1  nonMalignant   \n",
       "86                  1                  0                  0  nonMalignant   \n",
       "\n",
       "                   Patient_group  \n",
       "2196                     Sarcoma  \n",
       "1503  Non-Small-Cell Lung Cancer  \n",
       "689                Breast Cancer  \n",
       "1061       Asymptomatic controls  \n",
       "86         Asymptomatic controls  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove columns that are repeated or irrelevant\n",
    "df_y_merged.drop('Group', axis=1, inplace=True)\n",
    "df_y_merged.drop('lib.size', axis=1, inplace=True)\n",
    "df_y_merged.drop('classificationScoreCancer', axis=1, inplace=True)\n",
    "df_y_merged.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 2351 entries, 0 to 2350\n",
      "Data columns (total 10 columns):\n",
      " #   Column                        Non-Null Count  Dtype  \n",
      "---  ------                        --------------  -----  \n",
      " 0   Sample ID                     2351 non-null   object \n",
      " 1   Stage                         1372 non-null   object \n",
      " 2   Sex                           2331 non-null   object \n",
      " 3   Age                           2328 non-null   float64\n",
      " 4   Sample-supplying institution  2351 non-null   object \n",
      " 5   Training series               2351 non-null   int64  \n",
      " 6   Evaluation series             2351 non-null   int64  \n",
      " 7   Validation series             2351 non-null   int64  \n",
      " 8   Class_group                   2351 non-null   object \n",
      " 9   Patient_group                 2351 non-null   object \n",
      "dtypes: float64(1), int64(3), object(6)\n",
      "memory usage: 202.0+ KB\n"
     ]
    }
   ],
   "source": [
    "df_y_merged.info()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine DataFrames containing info about platelets and ground truth\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1-Vumc-HD-101-TR922', '2-Vumc-HD-103-TR923', '3-Vumc-HD-108-TR924', '4-Vumc-HD-127-TR925', '5-Vumc-HD-130-TR926', '6-Vumc-HD-142-TR927', '7-Vumc-HD-143-TR928', '8-Vumc-HD-144-TR929', '9-Vumc-HD-145-TR930', '11-Vumc-HD-149-TR932']\n",
      "2351\n",
      "['Vumc-HD-101-TR922', 'Vumc-HD-103-TR923', 'Vumc-HD-108-TR924', 'Vumc-HD-127-TR925', 'Vumc-HD-130-TR926', 'Vumc-HD-142-TR927', 'Vumc-HD-143-TR928', 'Vumc-HD-144-TR929', 'Vumc-HD-145-TR930', 'Vumc-HD-149-TR932']\n",
      "2351\n"
     ]
    }
   ],
   "source": [
    "# Check patients names\n",
    "patients_d = list(df_data.index)\n",
    "print(patients_d[:10])\n",
    "print(len(patients_d))\n",
    "\n",
    "patients_y = list(df_y_merged['Sample ID'])\n",
    "print(patients_y[:10])\n",
    "print(len(patients_y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of patients:  2351 Number of name coincidences 2348 For example:  1-Vumc-HD-101-TR922  and  Vumc-HD-101-TR922\n",
      "\n",
      "Cases where names are not simmilar: \n",
      "countMatrix.320-10 and Vumc-P-PH-001-010-TR1505\n",
      "countMatrix.327-34 and Vumc-P-PH-001-034-TR1512\n",
      "countMatrix.506-NKI-NSCLC-107-270 and NKI-NSCLC-107-270-TR725\n"
     ]
    }
   ],
   "source": [
    "compare_patients = [ x.find(y)!=-1 for (x,y) in zip(patients_d, patients_y)]\n",
    "print(\"Number of patients: \", len(compare_patients), \"Number of name coincidences\",  sum(compare_patients), \"For example: \", patients_d[0], \" and \", patients_y[0])\n",
    "\n",
    "print(\"\\nCases where names are not simmilar: \")\n",
    "different_patients =  [ x.find(y)==-1 for (x,y) in zip(patients_d, patients_y)]\n",
    "different_patients_index = [i for i, x in enumerate(different_patients) if x]\n",
    "for idx in different_patients_index:\n",
    "    print(patients_d[idx], \"and \", end='')\n",
    "    print(patients_y[idx])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We assume both databases have same list of patients and in the same order.\n",
    "# Therefore, we can join them\n",
    "\n",
    "df_y_merged.index = patients_d\n",
    "df = pd.merge(df_data, df_y_merged, left_index=True, right_index=True, how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2351, 5261)\n"
     ]
    }
   ],
   "source": [
    "# df.head(5)\n",
    "# display(df)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Stage', 'Sex', 'Age']\n",
      "979\n",
      "20\n",
      "23\n"
     ]
    }
   ],
   "source": [
    "# Check which columns have Nan values\n",
    "list_na_columns = df.columns[df.isna().any()].tolist()\n",
    "print(list_na_columns)\n",
    "\n",
    "# df['Age'].describe()\n",
    "# df['Age'].dtype       # float64\n",
    "# df.loc[df['Age'] == 18]\n",
    "\n",
    "# df['Age'].isnull().values.any()\n",
    "# df.isnull().values.any()\n",
    "\n",
    "# df['Stage'].isna().sum()\n",
    "# df['Stage'].describe()\n",
    "\n",
    "print(df['Stage'].isnull().sum())\n",
    "print(df['Sex'].isnull().sum())\n",
    "print(df['Age'].isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save merged dataFrame\n",
    "file='DataPatientsWithGT.csv'\n",
    "df.to_csv(file, sep=';', na_rep='n.a.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001461</th>\n",
       "      <th>ENSG00000001497</th>\n",
       "      <th>ENSG00000001629</th>\n",
       "      <th>ENSG00000001631</th>\n",
       "      <th>ENSG00000002330</th>\n",
       "      <th>ENSG00000002549</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample-supplying institution</th>\n",
       "      <th>Training series</th>\n",
       "      <th>Evaluation series</th>\n",
       "      <th>Validation series</th>\n",
       "      <th>Class_group</th>\n",
       "      <th>Patient_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1-Vumc-HD-101-TR922</th>\n",
       "      <td>3.937800</td>\n",
       "      <td>1.555228</td>\n",
       "      <td>7.445603</td>\n",
       "      <td>13.655275</td>\n",
       "      <td>11.121538</td>\n",
       "      <td>3.471206</td>\n",
       "      <td>5.986674</td>\n",
       "      <td>4.978068</td>\n",
       "      <td>46.477912</td>\n",
       "      <td>25.632179</td>\n",
       "      <td>...</td>\n",
       "      <td>Vumc-HD-101-TR922</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-Vumc-HD-103-TR923</th>\n",
       "      <td>67.841060</td>\n",
       "      <td>2.119756</td>\n",
       "      <td>113.660546</td>\n",
       "      <td>18.024223</td>\n",
       "      <td>2.863275</td>\n",
       "      <td>15.139871</td>\n",
       "      <td>3.036190</td>\n",
       "      <td>18.222689</td>\n",
       "      <td>40.741981</td>\n",
       "      <td>42.549333</td>\n",
       "      <td>...</td>\n",
       "      <td>Vumc-HD-103-TR923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>54.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-Vumc-HD-108-TR924</th>\n",
       "      <td>82.445588</td>\n",
       "      <td>7.368499</td>\n",
       "      <td>187.021192</td>\n",
       "      <td>9.728903</td>\n",
       "      <td>3.066767</td>\n",
       "      <td>7.518260</td>\n",
       "      <td>23.872422</td>\n",
       "      <td>11.552102</td>\n",
       "      <td>24.673084</td>\n",
       "      <td>66.978931</td>\n",
       "      <td>...</td>\n",
       "      <td>Vumc-HD-108-TR924</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>Institute 13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>nonMalignant</td>\n",
       "      <td>Asymptomatic controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 5261 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ENSG00000000419  ENSG00000000460  ENSG00000000938  \\\n",
       "1-Vumc-HD-101-TR922         3.937800         1.555228         7.445603   \n",
       "2-Vumc-HD-103-TR923        67.841060         2.119756       113.660546   \n",
       "3-Vumc-HD-108-TR924        82.445588         7.368499       187.021192   \n",
       "\n",
       "                     ENSG00000001036  ENSG00000001461  ENSG00000001497  \\\n",
       "1-Vumc-HD-101-TR922        13.655275        11.121538         3.471206   \n",
       "2-Vumc-HD-103-TR923        18.024223         2.863275        15.139871   \n",
       "3-Vumc-HD-108-TR924         9.728903         3.066767         7.518260   \n",
       "\n",
       "                     ENSG00000001629  ENSG00000001631  ENSG00000002330  \\\n",
       "1-Vumc-HD-101-TR922         5.986674         4.978068        46.477912   \n",
       "2-Vumc-HD-103-TR923         3.036190        18.222689        40.741981   \n",
       "3-Vumc-HD-108-TR924        23.872422        11.552102        24.673084   \n",
       "\n",
       "                     ENSG00000002549  ...          Sample ID  Stage  Sex  \\\n",
       "1-Vumc-HD-101-TR922        25.632179  ...  Vumc-HD-101-TR922    NaN    F   \n",
       "2-Vumc-HD-103-TR923        42.549333  ...  Vumc-HD-103-TR923    NaN    F   \n",
       "3-Vumc-HD-108-TR924        66.978931  ...  Vumc-HD-108-TR924    NaN    F   \n",
       "\n",
       "                      Age  Sample-supplying institution  Training series  \\\n",
       "1-Vumc-HD-101-TR922  35.0                  Institute 13                1   \n",
       "2-Vumc-HD-103-TR923  54.0                  Institute 13                0   \n",
       "3-Vumc-HD-108-TR924  56.0                  Institute 13                0   \n",
       "\n",
       "                     Evaluation series  Validation series   Class_group  \\\n",
       "1-Vumc-HD-101-TR922                  0                  0  nonMalignant   \n",
       "2-Vumc-HD-103-TR923                  1                  0  nonMalignant   \n",
       "3-Vumc-HD-108-TR924                  0                  1  nonMalignant   \n",
       "\n",
       "                             Patient_group  \n",
       "1-Vumc-HD-101-TR922  Asymptomatic controls  \n",
       "2-Vumc-HD-103-TR923  Asymptomatic controls  \n",
       "3-Vumc-HD-108-TR924  Asymptomatic controls  \n",
       "\n",
       "[3 rows x 5261 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.5 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7efbd634259f8e5f569f443ec65758432c9a08fec7ecc1e2b0020067d32b9947"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
